src/main.rs

```diff
@@ -10,51 +10,66 @@ mod helpers;
+    // information.
+
+    /// The role this application process should assume. The example in the template provides two
+    /// roles: server and client. The server echoes whatever message the clients send to it.
+    #[clap(value_enum, long)] // value_enum => parse as enum. long => "--role" instead of "-r".
+    role: Role, // This is a mandatory argument.
+
+    /// The server's network address. The server listens on this address. The client sends messages
+    /// to this address.
+    #[clap(long, value_parser = ipv4_resolve)] // value_parser => parse "ip:port" using ipv4_resolve
+    address: Option<SocketAddr>, // Since this is an Option<T>, it is an optional argument.
+
+    /// If specified, a graph representation of the Hydroflow flow used by the program will be
+    /// printed to the console in the specified format. This parameter can be removed if your
+    /// application doesn't need this functionality.
     #[clap(value_enum, long)]
-    role: Role,
-    // #[clap(long)]
-    #[clap(long, value_parser = ipv4_resolve)]
-    addr: Option<SocketAddr>,
-    // #[clap(long)]
-    #[clap(long, value_parser = ipv4_resolve)]
-    server_addr: Option<SocketAddr>,
-    #[clap(value_enum, long)]
-    graph: Option<GraphType>,
+    graph: Option<GraphType>, // This too, is an optional argument.
 }
 
 #[hydroflow::main]
+/// This is the main entry-point for both `Client` and `Server`.
 async fn main() {
-    // parse command line arguments
+    // Parse command line arguments
     let opts = Opts::parse();
-    // if no addr was provided, we ask the OS to assign a local port by passing in "localhost:0"
-    let addr = opts
-        .addr
-        .unwrap_or_else(|| ipv4_resolve("localhost:0").unwrap());
-
-    // allocate `outbound` sink and `inbound` stream
-    let (outbound, inbound, addr) = bind_udp_bytes(addr).await;
-    println!("Listening on {:?}", addr);
 
+    // Run the server or the client based on the role provided in the command-line arguments.
     match opts.role {
         Role::Server => {
-            run_server(outbound, inbound, opts).await;
+            run_server(opts).await;
         }
         Role::Client => {
-            run_client(outbound, inbound, opts).await;
+            run_client(opts).await;
         }
     }
 }
+
+/// A running application can assume one of these roles. The launched application process assumes
+/// one of these roles, based on the `--role` parameter passed in as a command line argument.
+#[derive(Clone, ValueEnum, Debug)]
+enum Role {
+    Client,
+    Server,
+}
+
+/// Optionally, the application can print a graph representation
+#[derive(Clone, ValueEnum, Debug)]
+pub enum GraphType {
+    Mermaid,
+    Dot,
+}
+
+/// The default server address & port on which the server listens for incoming messages. Clients
+/// send message to this address & port.
+pub const DEFAULT_SERVER_ADDRESS: &str = "localhost:54321";
```

Comment:
Wonder if it would be worth using `cargo template` to generate a random port each generation to avoid collisions

Reply:
Since the example is for running a single server, think this should be OK. If you feel strongly about this, I'd prefer to take it up separately.